src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php

```diff
@@ -48,45 +50,89 @@ public function updatePositionsAction(Request $request): Response
+        $productTaxonsPositions = $request->request->all('productTaxons');
 
-        $this->validateCsrfProtection($request, $configuration);
+        if (!$this->shouldProductsPositionsBeUpdated($request, $productTaxonsPositions)) {
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-        if ($this->shouldProductsPositionsBeUpdated($request, $productTaxons)) {
+        try {
+            $productTaxonsPositionsMap = $this->mapFromArrayToProductTaxonPositionMap($productTaxonsPositions);
+        } catch (\InvalidArgumentException $exception) {
             /** @var Session $session */
             $session = $request->getSession();
+            $session->getFlashBag()->add('error', $exception->getMessage());
 
-            foreach ($productTaxons as $id => $position) {
-                try {
-                    $this->updatePositions($position, $id);
-                } catch (\InvalidArgumentException $exception) {
-                    $session->getFlashBag()->add('error', $exception->getMessage());
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-                    return $this->redirectHandler->redirectToReferer($configuration);
-                }
-            }
+        /** @var PositionerInterface $positioner */
+        $positioner = $this->get(PositionerInterface::class);
 
+        foreach ($productTaxonsPositionsMap as $productTaxonPositionMap) {
+            $positioner->updatePosition(
+                $productTaxonPositionMap['productTaxon'],
+                $productTaxonPositionMap['newPosition'],
+                $this->getMaxPosition(),
+            );
             $this->manager->flush();
         }
 
         return $this->redirectHandler->redirectToReferer($configuration);
     }
 
+    /**
+     * @param array<int, string> $productTaxonPositions
+     *
+     * @return array<array{productTaxon: ProductTaxonInterface, newPosition: int}>
+     */
+    private function mapFromArrayToProductTaxonPositionMap(array $productTaxonPositions): array
+    {
+        /** @var PositionerInterface $positioner */
+        $positioner = $this->get(PositionerInterface::class);
+        $map = [];
+
+        foreach ($productTaxonPositions as $productTaxonId => $productTaxonPosition) {
+            if (!is_numeric($productTaxonPosition)) {
+                throw new \InvalidArgumentException(sprintf('The position "%s" is invalid.', $productTaxonPosition));
+            }
+
+            /** @var ProductTaxonInterface $productTaxon */
+            $productTaxon = $this->repository->find($productTaxonId);
+            $productTaxonPosition = (int) $productTaxonPosition;
+
+            if (!$positioner->hasPositionChanged($productTaxon, $productTaxonPosition)) {
+                continue;
+            }
+
+            $map[] = [
+                'productTaxon' => $productTaxon,
+                'newPosition' => $productTaxonPosition,
+            ];
```

Comment:
If I see correctly, you don't do anything else with this map, can't we instead of creating it and updating the positions in another foreach, do it here (especially that we have a positioner here)? Of course, you would have to adjust the name of the method a bit.

Reply:
I've checked this (and even started refactoring it 🙈), and I realized it has to be done in two separate foreach statements. 

In the first one we select values from db. In the second one, we update them. When I update the entity here, the listener is executed. When we start processing next product taxon it might have another position, due to the listener.